### PR TITLE
Implement bidirectional max-margin ranking loss

### PR DESCRIPTION
**Pull request by Naqu6**
Note: I've written up a doc on the vectorization of loss - happy to send it over if you're reviewing and want to see it.

Work done:
- Implement bidirectional max-margin ranking loss
- Add tests for metric

### CODE REVIEW

File: videoretrieval/metrics/loss.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    """Implementation of the Bidirectional max margin ranking loss.
+
+    Arguments:
+        video_embeddings: a tensor of dimension (batch size x embedding space
+            size) where the ith element is the embedding for the ith video in
+            the batch
+        text_embeddings: a tensor of dimension (batch size x embedding space
+            size) where the ith element is the embedding for the ith caption in
+            the batch
+        m: a hyper parameter
```
Comment (sbucak) [inline videoretrieval/metrics/loss.py:30]: From style guide: "avoid single character names, except for specifically allowed cases." What hyper parameter is this. Ideally name should explain it. The comment is also not very helpful here.

File: videoretrieval/metrics/loss.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    matching_similarities = tf.expand_dims(matching_similarities, 1)
+    matching_similarities = tf.tile(matching_similarities, [1, batch_size])
+
+    computed_similarities = tf.nn.relu(m + similarities - matching_similarities)
+    computed_similarities += tf.nn.relu(
+        m + similarities_transpose - matching_similarities)
+
+    computed_similarities = computed_similarities *  (1 - tf.eye(batch_size))
+
+    loss = tf.reduce_sum(computed_similarities) / (batch_size**2)
```
Comment (sbucak) [inline videoretrieval/metrics/loss.py:56]: is there a chance to have batch_size=0. Do we need to check that?

File: videoretrieval/tests/loss_tests.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            [-0.5547002,  0.8320503],
+            [0.0, 1.0],
+        ])
+
+        loss = bidirectional_max_margin_ranking_loss(mock_video_embeddings,
+            mock_text_embeddings, 1.0)
+
+        expected_value = 0.6779908
+
+        self.assertTrue(abs(loss.numpy() - expected_value) < self.error)
```
Comment (sbucak) [inline videoretrieval/tests/loss_tests.py:71]: nit: You can directly use the value of expected_value here just like you did in the previous test. This is the suggested way for C++ tests and probably valid for Python, too.

File: videoretrieval/tests/loss_tests.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+            [-1.0, 0.0],
+            [0, 1.0],
+            [-1.0, 1.0],
+            [0.7, 0.6],
+        ])
+
+        loss = bidirectional_max_margin_ranking_loss(mock_video_embeddings,
+            mock_text_embeddings, 1.5)
+
+        expected_value = 1.815000005
```
Comment (sbucak) [inline videoretrieval/tests/loss_tests.py:93]: ditto.

File: videoretrieval/metrics/loss.py (hunk 2 above)
Comment (Naqu6) [reply videoretrieval/metrics/loss.py:56]: There's no chance that will happen.

File: videoretrieval/tests/loss_tests.py (hunk 3 above)
Comment (Naqu6) [reply videoretrieval/tests/loss_tests.py:71]: Will do. Appreciate the feedback.

File: videoretrieval/metrics/loss.py (hunk 1 above)
Comment (Naqu6) [reply videoretrieval/metrics/loss.py:30]: Totally. Just updated!

**Comment by Naqu6** [comment]
Thanks!
